extractor/filesystem/language/javascript/packagelockjson/extractor.go

```diff
@@ -1,37 +1,236 @@
+	Name     string `json:"name"`
+	Version  string `json:"version"`
+	Resolved string `json:"resolved"`
+
+	Dependencies         map[string]string `json:"dependencies,omitempty"`
+	DevDependencies      map[string]string `json:"devDependencies,omitempty"`
+	OptionalDependencies map[string]string `json:"optionalDependencies,omitempty"`
+	PeerDependencies     map[string]string `json:"peerDependencies,omitempty"`
+
+	Dev         bool `json:"dev,omitempty"`
+	DevOptional bool `json:"devOptional,omitempty"`
+	Optional    bool `json:"optional,omitempty"`
+
+	Link bool `json:"link,omitempty"`
+}
+
+type npmLockfile struct {
+	Version int `json:"lockfileVersion"`
+	// npm v1- lockfiles use "dependencies"
+	Dependencies map[string]npmLockDependency `json:"dependencies,omitempty"`
+	// npm v2+ lockfiles use "packages"
+	Packages map[string]npmLockPackage `json:"packages,omitempty"`
+}
+
+type packageDetails struct {
+	Name      string
+	Version   string
+	Commit    string
+	DepGroups []string
+}
+
+const npmEcosystem string = "npm"
+
+type npmPackageDetailsMap map[string]packageDetails
+
+// mergeNpmDepsGroups handles merging the dependency groups of packages within the
+// NPM ecosystem, since they can appear multiple times in the same dependency tree
+//
+// the merge happens almost as you'd expect, except that if either given packages
+// belong to no groups, then that is the result since it indicates the package
+// is implicitly a production dependency.
+func mergeNpmDepsGroups(a, b packageDetails) []string {
+	// if either group includes no groups, then the package is in the "production" group
+	if len(a.DepGroups) == 0 || len(b.DepGroups) == 0 {
+		return nil
+	}
+
+	combined := make([]string, 0, len(a.DepGroups)+len(b.DepGroups))
+	combined = append(combined, a.DepGroups...)
+	combined = append(combined, b.DepGroups...)
+
+	slices.Sort(combined)
+
+	return slices.Compact(combined)
+}
+
+func (pdm npmPackageDetailsMap) add(key string, details packageDetails) {
+	existing, ok := pdm[key]
+
+	if ok {
+		details.DepGroups = mergeNpmDepsGroups(existing, details)
+	}
+
+	pdm[key] = details
+}
+
+func (dep npmLockDependency) depGroups() []string {
+	if dep.Dev && dep.Optional {
+		return []string{"dev", "optional"}
+	}
+	if dep.Dev {
+		return []string{"dev"}
+	}
+	if dep.Optional {
+		return []string{"optional"}
+	}
+
+	return nil
+}
+
+func parseNpmLockDependencies(dependencies map[string]npmLockDependency) map[string]packageDetails {
+	details := npmPackageDetailsMap{}
+
+	for name, detail := range dependencies {
+		if detail.Dependencies != nil {
+			nestedDeps := parseNpmLockDependencies(detail.Dependencies)
+			for k, v := range nestedDeps {
+				details.add(k, v)
+			}
+		}
+
+		version := detail.Version
+		finalVersion := version
+		commit := ""
+
+		// If the package is aliased, get the name and version
+		if strings.HasPrefix(detail.Version, "npm:") {
+			i := strings.LastIndex(detail.Version, "@")
```

Comment:
Add a comment that versions follow the format npm:name@version or something similar
Also I wonder if a regex would look nicer here.

Reply:
Hmm I don't think a regex is worth it here for just these 4 lines, especially since it splits the parsing logic (the regex string itself will be at the top of the file).